File: _ext/Past-Paper-Chatbot/scripts/retrieve.py
"""
Retrieval for Ask-AI over the Supabase REST API (PostgREST) — the same channel
the app uses for everything else, so it works against the firewalled Oracle
self-hosted Supabase from anywhere with HTTPS. No direct Postgres connection.

Semantic search calls the public.match_ask_ai_chunks RPC (migration 024); the
exact-term boost is a REST `ilike` select. Two modes (unchanged):
  paper_lookup — "which years / how often was X asked"
  general_qa   — "explain X" (grounding chunks for the LLM)

Every search is scoped to ONE level (olevel|alevel) so O- and A-level never mix,
and optionally to a subject.

CLI test (needs ASK_AI_SUPABASE_URL + ASK_AI_SUPABASE_SERVICE_KEY):
    python scripts/retrieve.py "which years was electrolysis asked" --level olevel
"""

import datetime
import os
import re
import sys
import threading
from pathlib import Path

from dotenv import load_dotenv
from sentence_transformers import SentenceTransformer
from supabase import create_client

PROJECT_DIR = Path(__file__).resolve().parent.parent
load_dotenv(PROJECT_DIR / ".env")

MODEL_NAME = os.environ.get("ASK_AI_EMBED_MODEL", "BAAI/bge-base-en-v1.5")
# bge retrieval convention: QUERIES get this instruction; the passages in the
# index do not (the build job embeds them raw). Keep in sync with the build.
QUERY_INSTRUCTION = "Represent this sentence for searching relevant passages: "


class VectorStoreUnavailable(RuntimeError):
    """The REST index is unreachable or not built (migration 024 not applied) —
    the HTTP layer turns this into a clear 503, and the container still boots."""


PAPER_LOOKUP_PATTERNS = [
    r"\bwhich years?\b", r"\bwhat years?\b", r"\bhow many times\b", r"\bhow often\b",
    r"\bhas .* (been )?asked\b", r"\bwas .* asked\b", r"\bpast papers?\b",
    r"\bprevious (years?|papers?|exams?)\b", r"\bshow me questions?\b",
    r"\bfind questions?\b", r"\blist questions?\b",
]
PAPER_LOOKUP_RE = re.compile("|".join(PAPER_LOOKUP_PATTERNS), re.IGNORECASE)

YEAR_LIMIT_RE = re.compile(r"\b(?:last|past|previous|recent)\s+(\d{1,2})\s+years?\b", re.IGNORECASE)
DEFAULT_YEAR_LIMIT = 5


def requested_year_limit(query: str) -> int:
    m = YEAR_LIMIT_RE.search(query)
    return int(m.group(1)) if m else DEFAULT_YEAR_LIMIT


# Union of O- and A-level subject names, for auto-detecting a subject from free
# text when the UI didn't pin one. Longest first so "Business Studies" /
# "Computer Science" match before a shorter substring would.
SUBJECTS = [
    "Accounting", "Additional Maths", "Art and Design", "Biology", "Business Studies",
    "Business", "Chemistry", "Commerce", "Computer Science", "Economics",
    "English General Paper", "English Language", "English", "Environmental Management",
    "Further Mathematics", "Geography", "Global Perspectives", "History",
    "Information Technology", "Islamiyat", "Law", "Literature in English",
    "Mathematics", "Pakistan Studies", "Physics", "Psychology", "Religious Studies",
    "Sociology", "Statistics",
]
_SUBJECT_RES = sorted(
    ((s, re.compile(r"\b" + re.escape(s) + r"\b", re.IGNORECASE)) for s in SUBJECTS),
    key=lambda pair: -len(pair[0]),
)

ACRONYM_RE = re.compile(r"\b[A-Z]{2,6}\b")

GENERIC_QUERY_WORDS = {
    "in", "the", "a", "an", "was", "were", "is", "are", "asked", "for", "of",
    "on", "about", "come", "up", "to", "this", "that", "topic", "questions",
    "question", "how", "many", "times", "did", "has", "have", "had", "been",
    "being", "which", "what", "years", "year", "show", "find", "list", "past",
    "papers", "paper", "and", "related", "does", "do",
}


def detect_subject(query: str):
    for name, pattern in _SUBJECT_RES:
        if pattern.search(query):
            return name
    return None


def extract_core_topic(query: str, subject=None) -> str:
    text = YEAR_LIMIT_RE.sub(" ", query)
    if subject:
        text = re.sub(re.escape(subject), " ", text, flags=re.IGNORECASE)
    text = re.sub(r"[?.!,]", " ", text)
    words = [w for w in text.split() if w.lower() not in GENERIC_QUERY_WORDS]
    return " ".join(words).strip()


def normalize_level(level):
    """Canonicalize to the value stored in questions.level ('olevel'|'alevel').
    Accepts 'O'/'A', 'O Level', 'a_level', etc. None -> no level filter."""
    if not level:
        return None
    s = re.sub(r"[^a-z]", "", str(level).lower())
    if s.startswith("o"):
        return "olevel"
    if s.startswith("a"):
        return "alevel"
    return None


# --- shared, lazily-initialised singletons (model + REST client) --------------
_model = None
_model_lock = threading.Lock()
_sb = None
_sb_lock = threading.Lock()


def get_model() -> SentenceTransformer:
    global _model
    if _model is None:
        with _model_lock:
            if _model is None:
                _model = SentenceTransformer(MODEL_NAME)
    return _model


def get_sb():
    global _sb
    if _sb is None:
        url = (os.environ.get("ASK_AI_SUPABASE_URL") or os.environ.get("SUPABASE_URL") or "").strip()
        key = (os.environ.get("ASK_AI_SUPABASE_SERVICE_KEY")
               or os.environ.get("SUPABASE_SERVICE_ROLE_KEY")
               or os.environ.get("SUPABASE_KEY") or "").strip()
        if not url or not key:
            raise VectorStoreUnavailable(
                "Supabase REST not configured (ASK_AI_SUPABASE_URL / ASK_AI_SUPABASE_SERVICE_KEY)."
            )
        with _sb_lock:
            if _sb is None:
                _sb = create_client(url, key)
    return _sb


_SELECT_COLS = ("question_id,level,subject,type,exam_year,session,paper,variant,"
                "question_number,topic,content")


def _to_hit(r: dict, verified: bool) -> dict:
    """Map an ask_ai_chunks row to the hit shape generate_answer.py expects
    (exam_year -> year). No source PDF page from the DB, so source_file/page are
    None — citations then show the paper reference without a page image."""
    meta = {
        "subject": r.get("subject"),
        "year": r.get("exam_year"),
        "session": r.get("session"),
        "paper": r.get("paper"),
        "variant": r.get("variant"),
        "question_number": r.get("question_number"),
        "topic": r.get("topic"),
        "level": r.get("level"),
        "type": r.get("type"),
        "source_file": None,
        "page": None,
    }
    return {
        "text": r.get("content") or "",
        "metadata": meta,
        "distance": float(r.get("distance") or 0.0),
        "verified": verified,
    }


class Retriever:
    def __init__(self):
        self.model = get_model()

    def classify_intent(self, query: str) -> str:
        return "paper_lookup" if PAPER_LOOKUP_RE.search(query) else "general_qa"

    def search(self, query: str, subject=None, level=None, top_k: int = 10, core_topic=None):
        vec = self.model.encode([QUERY_INSTRUCTION + query], normalize_embeddings=True)[0]
        # pgvector text form — PostgREST casts this string to vector for the RPC.
        qvec = "[" + ",".join(f"{float(x):.6f}" for x in vec) + "]"
        level = normalize_level(level)
        sb = get_sb()
        try:
            # Semantic search via the pgvector RPC (always scoped by level).
            semantic = sb.rpc("match_ask_ai_chunks", {
                "query_embedding": qvec,
                "match_level": level,
                "match_subject": subject,
                "match_count": top_k,
            }).execute().data or []

            # Exact-term boost: acronyms + core topic, so a genuine match just
            # outside top-k (or an acronym embeddings rank poorly) isn't dropped.
            terms = set(ACRONYM_RE.findall(query))
            if core_topic and len(core_topic) >= 4:
                terms.add(core_topic)
            verified_ids = set()
            exact = []
            for term in terms:
                q = sb.table("ask_ai_chunks").select(_SELECT_COLS).ilike("content", f"*{term}*").limit(200)
                if level:
                    q = q.eq("level", level)
                if subject:
                    q = q.eq("subject", subject)
                for r in (q.execute().data or []):
                    verified_ids.add(r["question_id"])
                    exact.append(r)
        except VectorStoreUnavailable:
            raise
        except Exception as e:
            raise VectorStoreUnavailable(
                f"Ask-AI index query failed (is migration 024 applied and the index built?): {e}"
            ) from e

        by_id: dict = {}
        for r in semantic:
            by_id[r["question_id"]] = _to_hit(r, verified=False)
        for r in exact:
            if r["question_id"] not in by_id:
                by_id[r["question_id"]] = _to_hit(r, verified=True)
        for qid in verified_ids:
            if qid in by_id:
                by_id[qid]["verified"] = True

        hits = sorted(by_id.values(), key=lambda h: h["distance"])
        verified_hits = [h for h in hits if h["verified"]]
        other_hits = [h for h in hits if not h["verified"]]
        return verified_hits + other_hits[: max(0, top_k - len(verified_hits))]

    def paper_lookup_summary(self, hits, year_limit=None):
        verified_hits = [h for h in hits if h.get("verified")]
        candidates = verified_hits if verified_hits else hits
        seen = set()
        occurrences = []
        for h in candidates:
            m = h["metadata"]
            if m.get("question_number") is None:
                continue
            key = (m.get("subject"), m.get("year"), m.get("session"), m.get("paper"),
                   m.get("variant"), m.get("question_number"))
            if key in seen:
                continue
            seen.add(key)
            occurrences.append({**m, "question_text": h["text"]})
        if year_limit is not None and occurrences:
            this_year = datetime.date.today().year
            cutoff_year = this_year - year_limit + 1
            occurrences = [m for m in occurrences if (m.get("year") or 0) >= cutoff_year]
        occurrences.sort(key=lambda m: (m.get("year") or 0, m.get("session") or ""))
        return occurrences

    def route(self, query: str, subject=None, level=None, top_k: int = 10):
        intent = self.classify_intent(query)
        if subject is None:
            subject = detect_subject(query)
        search_query = YEAR_LIMIT_RE.sub("", query).strip()
        core_topic = extract_core_topic(query, subject)
        hits = self.search(search_query, subject=subject, level=level, top_k=top_k, core_topic=core_topic)
        result = {"intent": intent, "hits": hits}
        if intent == "paper_lookup":
            year_limit = requested_year_limit(query)
            result["year_limit"] = year_limit
            result["occurrences"] = self.paper_lookup_summary(hits, year_limit=year_limit)
        return result


def main():
    args = sys.argv[1:]
    level = None
    if "--level" in args:
        i = args.index("--level")
        level = args[i + 1] if i + 1 < len(args) else None
        del args[i:i + 2]
    query = args[0] if args else "which years was depreciation asked in accounting"

    r = Retriever()
    result = r.route(query, level=level)
    print(f"Query: {query!r}  (level={level})")
    print(f"Detected intent: {result['intent']}\n")
    if result["intent"] == "paper_lookup":
        print("Occurrences found:")
        for m in result["occurrences"]:
            print(f"  - {m.get('subject')} {m.get('year')} {m.get('session')} "
                  f"Paper {m.get('paper')} Variant {m.get('variant')} Q{m.get('question_number')}")
        print()
    print("Top retrieved chunks:")
    for h in result["hits"][:5]:
        m = h["metadata"]
        print(f"--- distance={h['distance']:.4f} | {m.get('subject')} {m.get('year')} "
              f"{m.get('session')} Paper {m.get('paper')} Q{m.get('question_number')} ---")
        print(h["text"][:200].replace("\n", " "))
        print()


if __name__ == "__main__":
    main()

File: src/lib/askai/retrieve.ts
/**
 * Retrieval for Ask-AI, in-process over the Supabase REST client (pgvector via
 * the match_ask_ai_chunks RPC + a keyword pass over ask_ai_chunks.content).
 *
 * Every search is scoped to ONE level (olevel|alevel) so O- and A-level results
 * never mix, and optionally to a subject. The planner (planner.ts) decides the
 * search phrasings and keywords; rank.ts sorts the candidates into tiers.
 */
import { supabase } from '../supabase';
import { embedQuery } from './embed';

export type Level = 'olevel' | 'alevel';

export interface HitMeta {
  subject: string | null; year: number | null; session: string | null;
  paper: string | null; variant: string | null; question_number: string | null;
  topic: string | null; level: string | null; type: string | null;
}
export interface Hit {
  id: string;
  text: string;
  metadata: HitMeta;
  /** cosine distance from the query (lower = closer); keyword-only rows get a neutral 0.6 */
  distance: number;
  /** true when the row literally contains one of the planner's keywords */
  verified: boolean;
}

// Subjects present in the index, per level (probed from ask_ai_chunks). The
// planner may only pick from these, so a hallucinated subject can't zero a search.
export const SUBJECTS_BY_LEVEL: Record<Level, string[]> = {
  olevel: [
    'Accounting', 'Art and Design', 'Biology', 'Business Studies', 'Chemistry', 'Commerce',
    'Computer Science', 'Economics', 'English Language', 'Environmental Management', 'Geography',
    'History', 'Islamiyat', 'Mathematics', 'Pakistan Studies', 'Physics', 'Religious Studies',
    'Sociology', 'Statistics',
  ],
  alevel: [
    'Accounting', 'Biology', 'Business', 'Chemistry', 'Computer Science', 'Economics',
    'English General Paper', 'English Language', 'Further Mathematics', 'Geography',
    'Global Perspectives', 'History', 'Information Technology', 'Law', 'Literature in English',
    'Mathematics', 'Physics', 'Psychology', 'Sociology',
  ],
};
export const ALL_SUBJECTS = [...new Set([...SUBJECTS_BY_LEVEL.olevel, ...SUBJECTS_BY_LEVEL.alevel])].sort();

// Common student shorthand → canonical index subject (level-specific ones are
// resolved in resolveSubject).
const SUBJECT_ALIASES: Record<string, string> = {
  maths: 'Mathematics', math: 'Mathematics', mathematics: 'Mathematics',
  'further maths': 'Further Mathematics', 'further mathematics': 'Further Mathematics',
  bio: 'Biology', biology: 'Biology', chem: 'Chemistry', chemistry: 'Chemistry',
  phy: 'Physics', phys: 'Physics', physics: 'Physics',
  eco: 'Economics', econ: 'Economics', economics: 'Economics',
  accounts: 'Accounting', accounting: 'Accounting',
  'pak studies': 'Pakistan Studies', 'pakistan studies': 'Pakistan Studies', pst: 'Pakistan Studies',
  islamiat: 'Islamiyat', islamiyat: 'Islamiyat',
  cs: 'Computer Science', computer: 'Computer Science', 'computer science': 'Computer Science', computing: 'Computer Science',
  business: 'Business', 'business studies': 'Business Studies',
  english: 'English Language', 'english language': 'English Language',
  gp: 'English General Paper', 'general paper': 'English General Paper', 'english general paper': 'English General Paper',
  lit: 'Literature in English', literature: 'Literature in English', 'english literature': 'Literature in English', 'literature in english': 'Literature in English',
  it: 'Information Technology', ict: 'Information Technology', 'information technology': 'Information Technology',
  env: 'Environmental Management', evm: 'Environmental Management', 'environmental management': 'Environmental Management',
  stats: 'Statistics', statistics: 'Statistics', socio: 'Sociology', sociology: 'Sociology',
  psych: 'Psychology', psychology: 'Psychology', rs: 'Religious Studies', 'religious studies': 'Religious Studies',
  geo: 'Geography', geography: 'Geography', law: 'Law', history: 'History', commerce: 'Commerce',
  art: 'Art and Design', 'art and design': 'Art and Design', 'global perspectives': 'Global Perspectives',
};

export function normalizeLevel(level?: string | null): Level | null {
  if (!level) return null;
  const s = String(level).toLowerCase().replace(/[^a-z]/g, '');
  if (s.startsWith('o')) return 'olevel';
  if (s.startsWith('a')) return 'alevel';
  return null;
}

/** Map any subject spelling ("Chem", "Business (A Level)") to the index name at this level, or null. */
export function resolveSubject(raw: string | null | undefined, level: Level): string | null {
  if (!raw) return null;
  const s = String(raw).replace(/\s*\((?:A|O)[ -]?Levels?\)\s*$/i, '').trim();
  if (!s) return null;
  const list = SUBJECTS_BY_LEVEL[level];
  const lower = s.toLowerCase();
  let canonical = list.find((n) => n.toLowerCase() === lower) || SUBJECT_ALIASES[lower] || null;
  if (!canonical) canonical = list.find((n) => n.toLowerCase().startsWith(lower)) || null;
  if (!canonical) return null;
  // Business is "Business Studies" at O Level and "Business" at A Level.
  if (/^business/i.test(canonical)) return level === 'olevel' ? 'Business Studies' : 'Business';
  return list.includes(canonical) ? canonical : null;
}

// Subject mentions in free text, longest first so "Business Studies" wins over "Business".
const SUBJECT_MENTIONS = Object.keys(SUBJECT_ALIASES)
  .concat(ALL_SUBJECTS.map((s) => s.toLowerCase()))
  .filter((v, i, arr) => arr.indexOf(v) === i && v.length > 2)
  .sort((a, b) => b.length - a.length)
  .map((name) => [name, new RegExp('\\b' + name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&') + '\\b', 'i')] as const);

export function detectSubject(query: string, level: Level): string | null {
  for (const [name, re] of SUBJECT_MENTIONS) {
    if (re.test(query)) {
      const resolved = resolveSubject(name, level);
      if (resolved) return resolved;
    }
  }
  return null;
}

// ---- heuristics shared with the planner's no-LLM fallback ----------------------

const PAPER_LOOKUP_RE = new RegExp([
  '\\bwhich years?\\b', '\\bwhat years?\\b', '\\bhow many times\\b', '\\bhow often\\b',
  '\\bhas .* (been )?asked\\b', '\\bwas .* asked\\b', '\\bpast papers?\\b',
  '\\bprevious (years?|papers?|exams?)\\b', '\\bshow me questions?\\b',
  '\\bfind questions?\\b', '\\blist questions?\\b', '\\bwhere (did|does|has) .* (come|appear)\\b',
].join('|'), 'i');
export const looksLikePaperLookup = (q: string) => PAPER_LOOKUP_RE.test(q);

const YEAR_LIMIT_RE = /\b(?:last|past|previous|recent)\s+(\d{1,2})\s+years?\b/i;
export function requestedYearFrom(query: string): number | null {
  const m = query.match(YEAR_LIMIT_RE);
  return m ? new Date().getFullYear() - parseInt(m[1], 10) + 1 : null;
}
export const stripYearPhrase = (q: string) => q.replace(YEAR_LIMIT_RE, ' ').replace(/\s+/g, ' ').trim();

const ACRONYM_RE = /\b[A-Z]{2,6}\b/g;
const GENERIC_WORDS = new Set([
  'in', 'the', 'a', 'an', 'was', 'were', 'is', 'are', 'asked', 'for', 'of', 'on', 'about', 'come',
  'up', 'to', 'this', 'that', 'topic', 'questions', 'question', 'how', 'many', 'times', 'did',
  'has', 'have', 'had', 'been', 'being', 'which', 'what', 'years', 'year', 'show', 'find', 'list',
  'past', 'papers', 'paper', 'and', 'related', 'does', 'do', 'me', 'explain', 'give', 'please',
  'with', 'from', 'level', 'exam', 'exams', 'mcq', 'mcqs',
]);

/** Query minus subject/years/filler — used as a keyword when the planner is unavailable. */
export function extractCoreTopic(query: string, subject: string | null): string {
  let text = stripYearPhrase(query);
  if (subject) text = text.replace(new RegExp(subject.replace(/[.*+?^${}()|[\]\\]/g, '\\$&'), 'ig'), ' ');
  text = text.replace(/[?.!,;:"']/g, ' ');
  return text.split(/\s+/).filter((w) => w && !GENERIC_WORDS.has(w.toLowerCase())).join(' ').trim();
}

export function heuristicKeywords(query: string, subject: string | null): string[] {
  const terms = new Set<string>(query.match(ACRONYM_RE) || []);
  const core = extractCoreTopic(query, subject);
  if (core && core.length >= 4 && core.split(' ').length <= 3) terms.add(core);
  return [...terms].slice(0, 4);
}

// ---- search --------------------------------------------------------------------

const SELECT_COLS =
  'question_id,level,subject,type,exam_year,session,paper,variant,question_number,topic,content';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
function toHit(r: any, verified: boolean, distance?: number): Hit {
  return {
    id: String(r.question_id),
    text: r.content || '',
    metadata: {
      subject: r.subject ?? null, year: r.exam_year ?? null, session: r.session ?? null,
      paper: r.paper ?? null, variant: r.variant ?? null, question_number: r.question_number ?? null,
      topic: r.topic ?? null, level: r.level ?? null, type: r.type ?? null,
    },
    distance: distance ?? Number(r.distance ?? 0.6),
    verified,
  };
}

export class AskAiIndexUnavailable extends Error {}

export interface SearchOptions {
  level: Level;
  subject?: string | null;
  topK?: number;
  /** exact terms a matching question would contain — boosts semantic hits and pulls in literal mentions */
  keywords?: string[];
  questionType?: 'mcq' | 'structured' | null;
}

const KEYWORD_CLEAN_RE = /[^\p{L}\p{N} +\-./]/gu;

/** One semantic search + keyword pass, level-scoped, returning up to topK hits (closest first). */
export async function search(query: string, opts: SearchOptions): Promise<Hit[]> {
  const topK = opts.topK ?? 12;
  const qvec = await embedQuery(query);

  const rpc = await supabase.rpc('match_ask_ai_chunks', {
    query_embedding: qvec, match_level: opts.level, match_subject: opts.subject ?? null,
    match_count: Math.max(topK * 3, 30),
  });
  if (rpc.error) {
    throw new AskAiIndexUnavailable(
      `Ask-AI search failed (is migration 024 applied and the index built?): ${rpc.error.message}`,
    );
  }
  const byId = new Map<string, Hit>();
  for (const r of (rpc.data as unknown[]) || []) {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const hit = toHit(r as any, false);
    byId.set(hit.id, hit);
  }

  // Keyword pass: semantic hits that literally contain a term are "verified" and
  // nudged up; literal mentions the embedding missed are appended (most recent
  // first) so a genuine occurrence is never dropped.
  const terms = [...new Set((opts.keywords || [])
    .map((t) => t.replace(KEYWORD_CLEAN_RE, ' ').replace(/\s+/g, ' ').trim())
    .filter((t) => t.length >= 3))].slice(0, 4);
  const keywordIds = new Set<string>();
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const literalOnly = new Map<string, any>();
  await Promise.all(terms.map(async (term) => {
    let q = supabase.from('ask_ai_chunks').select(SELECT_COLS)
      .eq('level', opts.level).ilike('content', `*${term}*`)
      .order('exam_year', { ascending: false }).limit(120);
    if (opts.subject) q = q.eq('subject', opts.subject);
    if (opts.questionType) q = q.eq('type', opts.questionType);
    const { data, error } = await q;
    if (error) { console.warn('[askai] keyword pass failed:', error.message); return; }
    for (const r of (data as unknown[]) || []) {
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      const row = r as any;
      keywordIds.add(String(row.question_id));
      if (!byId.has(String(row.question_id))) literalOnly.set(String(row.question_id), row);
    }
  }));
  for (const id of keywordIds) {
    const h = byId.get(id);
    if (h) { h.verified = true; h.distance = Math.max(0, h.distance - 0.04); }
  }

  let ranked = [...byId.values()].sort((a, b) => a.distance - b.distance);
  if (opts.questionType) {
    // Prefer the requested type, but don't empty the result set over it.
    const typed = ranked.filter((h) => h.metadata.type === opts.questionType);
    if (typed.length >= Math.min(4, topK)) ranked = typed;
  }
  const tail = [...literalOnly.values()]
    .sort((a, b) => (b.exam_year || 0) - (a.exam_year || 0))
    .slice(0, Math.max(4, Math.floor(topK / 3)))
    .map((r) => toHit(r, true, 0.6));
  return [...ranked, ...tail].slice(0, topK);
}

/** Run several phrasings of the same need and merge (closest distance wins). */
export async function searchMany(queries: string[], opts: SearchOptions): Promise<Hit[]> {
  const topK = opts.topK ?? 12;
  const phrasings = [...new Set(queries.map((q) => q.trim()).filter(Boolean))].slice(0, 3);
  const results = await Promise.all(phrasings.map((q) => search(q, { ...opts, topK })));
  const merged = new Map<string, Hit>();
  for (const hits of results) {
    for (const h of hits) {
      const prev = merged.get(h.id);
      if (!prev) merged.set(h.id, { ...h });
      else { prev.distance = Math.min(prev.distance, h.distance); prev.verified = prev.verified || h.verified; }
    }
  }
  return [...merged.values()].sort((a, b) => a.distance - b.distance).slice(0, topK);
}

export function filterYears(hits: Hit[], from: number | null, to: number | null): Hit[] {
  if (from == null && to == null) return hits;
  return hits.filter((h) => {
    const y = h.metadata.year || 0;
    return (from == null || y >= from) && (to == null || y <= to);
  });
}

// ---- presentation helpers ------------------------------------------------------

export const prettySession = (s: string | null) => (s || '').replace(/_/g, '/');
export const prettyPaper = (p: string | null) => (p || '').replace(/_/g, ' ');
export function prettyVariant(v: string | null): string {
  if (!v) return '';
  const s = String(v).trim();
  return /^\d+$/.test(s) ? `Variant ${s}` : s.replace(/_/g, ' ');
}

/** "Chemistry 2023 May/June Paper 2 Variant 1 Q8" */
export function refLabel(m: HitMeta): string {
  return [m.subject, m.year, prettySession(m.session), prettyPaper(m.paper), prettyVariant(m.variant),
    m.question_number ? `Q${m.question_number}` : '']
    .filter(Boolean).join(' ');
}

export function dedupeByPaperQuestion(hits: Hit[]): Hit[] {
  const seen = new Set<string>();
  return hits.filter((h) => {
    const m = h.metadata;
    const key = [m.subject, m.year, m.session, m.paper, m.variant, m.question_number].join('|');
    if (seen.has(key)) return false;
    seen.add(key);
    return true;
  });
}
